### PR TITLE
Update callTracer and fastCallTracer to be more descriptive for SELFDESTRUCT operation.

### LINKED ISSUE #1335: Need to update callTracer and fastCallTracer to be more descriptive for SELFDESTRUCT
Added the fields `to, from, gas, gasUsed, input, output, to, and value` to the tracer when the op is `SELFDESTRUCT`. 
This lets us figure out how much KLAY was self destructed and helps with accounting.

Currently Klaytn tracer trace SELFDESTRUCT call as below.
```
{
  result: {
    calls: [{
        type: "SELFDESTRUCT"
    }],
    from: "0x226b9adf614da779196a7dab71a41b267163ea27",
    gas: "0x6fb8",
    gasUsed: "0x175c",
    input: "0x43d726d6",
    output: "0x",
    time: "53.791µs",
    to: "0xd1dd0db34e7ca8aaaeffe1f22e4ad98d2c9b1fdc",
    type: "CALL",
    value: "0x0"
  },
  txHash: "0xc62323d6c5a165f5992a5fd4cd754fbd01c4005a7f4fdb4f695ccc526e2b4aa0"
}
```

But Go Ethereum tracer trace SELFDESTRUCT call as below.
Using legacy javascript call tracer (which is Klaytn's call tracer).

```
> callResult = debug.traceBlockByHash("0x70d3ac44196eb7c831e0fb91db6d42da311f36529a81be6aaabbc81c36e5ab18", {tracer: "callTracerLegacy"})
> callResult[0]
{
  result: {
    calls: [{
        from: "0xbe3892d33620be5aca8c75d39e7401871194d290",
        to: "0xca7a99380131e6c76cfa622396347107aeedca2d",
        type: "SELFDESTRUCT",
        value: "0x10f0cf064dd59200000"
    }],
    from: "0xca7a99380131e6c76cfa622396347107aeedca2d",
    gas: "0x7108",
    gasUsed: "0x1c70",
    input: "0x43d726d6",
    output: "0x",
    time: "619.583µs",
    to: "0xbe3892d33620be5aca8c75d39e7401871194d290",
    type: "CALL",
    value: "0x0"
  }
}
```

Using native call tracer.
```
> callResult = debug.traceBlockByHash("0x70d3ac44196eb7c831e0fb91db6d42da311f36529a81be6aaabbc81c36e5ab18", {tracer: "callTracer"})
[{
    result: {
      calls: [{...}],
      from: "0xca7a99380131e6c76cfa622396347107aeedca2d",
      gas: "0x7108",
      gasUsed: "0x1c70",
      input: "0x43d726d6",
      output: "0x",
      to: "0xbe3892d33620be5aca8c75d39e7401871194d290",
      type: "CALL",
      value: "0x0"
    }
}]
> callResult[0]
{
  result: {
    calls: [{
        from: "0xbe3892d33620be5aca8c75d39e7401871194d290",
        gas: "0x0",
        gasUsed: "0x0",
        input: "0x",
        output: "0x",
        to: "0xca7a99380131e6c76cfa622396347107aeedca2d",
        type: "SELFDESTRUCT",
        value: "0x10f0cf064dd59200000"
    }],
    from: "0xca7a99380131e6c76cfa622396347107aeedca2d",
    gas: "0x7108",
    gasUsed: "0x1c70",
    input: "0x43d726d6",
    output: "0x",
    to: "0xbe3892d33620be5aca8c75d39e7401871194d290",
    type: "CALL",
    value: "0x0"
  }
}
```

Related Geth PR is https://github.com/ethereum/go-ethereum/pull/21549

cc. @jimni1222 @jeongkyun-oh

### PR DESCRIPTION
**Pull request by aeharvlee**
## Proposed changes

Both tracers now are more descriptive for SELFDESTRUCT operation.
This patch applied https://github.com/ethereum/go-ethereum/pull/21549 into Klaytn.

Closes #1335 

## How to test this PR

**1. Deploy the bytecode**: `0x608060405234801561001057600080fd5b507380fe91d59adabeb5a0e5f40edc1ef692025534ae6000806101000a81548173ffffffffffffffffffffffffffffffffffffffff021916908373ffffffffffffffffffffffffffffffffffffffff160217905550610145806100746000396000f3fe6080604052600436106100385760003560e01c80632e64cec11461004457806343d726d61461006f5780636057361d146100865761003f565b3661003f57005b600080fd5b34801561005057600080fd5b506100596100c1565b6040518082815260200191505060405180910390f35b34801561007b57600080fd5b506100846100cb565b005b34801561009257600080fd5b506100bf600480360360208110156100a957600080fd5b8101908080359060200190929190505050610105565b005b6000600154905090565b6000809054906101000a900473ffffffffffffffffffffffffffffffffffffffff1673ffffffffffffffffffffffffffffffffffffffff16ff5b806001819055505056fea26469706673582212201fdb1fc86dd3f982f18eb347791bcba28ffd0786b11d4e62276a2be3e8796c9a64736f6c63430006010033` to your own test-net.

You can get bytecode by compiling below solidity code.
```sol
pragma solidity >0.6.0;

contract Storage {
    address payable private owner;
    uint256 number;
    constructor() public {
        owner = 0x80fE91D59ADaBEb5a0e5F40edC1Ef692025534AE;
    }
    function store(uint256 num) public {
        number = num;
    }

    function retrieve() public view returns (uint256){
        return number;
    }

    receive() external payable {
        // React to receiving ether
    }

    function close() public {
        selfdestruct(owner);
    }
}
```

**Send transaction executing `close()` and get the block number including that tx**
ABI of close is `0x43d726d6`, so you can use this in input field of tx.

### As-Is
```js
> fastCallResult = debug.traceBlockByHash("0x19f8522fe7e879899b6c53fefd9fb1b5b566d90c4f3579bad35dd5e6a8a42e73", {tracer: "fastCallTracer"})
# Below result is same when using callTracer.
> fastCallResult[0]
{
  result: {
    calls: [{
        type: "SELFDESTRUCT"
    }],
    from: "0x650913630bf1a0f9d4ea90290463a6c5f7f023d1",
    gas: "0x74d88",
    gasUsed: "0x7904",
    input: "0x43d726d6",
    output: "0x",
    time: "1.867375709s",
    to: "0xf360ade8598e37a986b1c7436d40e18a23976574",
    type: "CALL",
    value: "0x0"
  },
  txHash: "0xb00d7cb3a0126a008cf67ae66719bf7c2bcaaed1a6df39010d173db511a860d7"
}
```

### To-Be (This patch is merged)
```js
> fastCallResult = debug.traceBlockByHash("0x19f8522fe7e879899b6c53fefd9fb1b5b566d90c4f3579bad35dd5e6a8a42e73", {tracer: "fastCallTracer"})
# Same result when using callTracer as option.
> fastCallResult[0]
{
  result: {
    calls: [{
        from: "0xf360ade8598e37a986b1c7436d40e18a23976574",
        to: "0x80fe91d59adabeb5a0e5f40edc1ef692025534ae",
        type: "SELFDESTRUCT",
        value: "0x2b5e3af16b1880000"
    }],
    from: "0x650913630bf1a0f9d4ea90290463a6c5f7f023d1",
    gas: "0x74d88",
    gasUsed: "0x7904",
    input: "0x43d726d6",
    output: "0x",
    time: "79.708µs",
    to: "0xf360ade8598e37a986b1c7436d40e18a23976574",
    type: "CALL",
    value: "0x0"
  },
  txHash: "0xb00d7cb3a0126a008cf67ae66719bf7c2bcaaed1a6df39010d173db511a860d7"
}
```

Above result formats are same with Geth's callTracerLegacy (Javascript Call Tracer of Geth).
```js
> callTracerResult = debug.traceBlockByHash("0x70d3ac44196eb7c831e0fb91db6d42da311f36529a81be6aaabbc81c36e5ab18", {tracer: "callTracerLegacy"})
> callTracerResult[0]
{
  result: {
    calls: [{
        from: "0xbe3892d33620be5aca8c75d39e7401871194d290",
        to: "0xca7a99380131e6c76cfa622396347107aeedca2d",
        type: "SELFDESTRUCT",
        value: "0x10f0cf064dd59200000"
    }],
    from: "0xca7a99380131e6c76cfa622396347107aeedca2d",
    gas: "0x7108",
    gasUsed: "0x1c70",
    input: "0x43d726d6",
    output: "0x",
    time: "654µs",
    to: "0xbe3892d33620be5aca8c75d39e7401871194d290",
    type: "CALL",
    value: "0x0"
  }
}
```

## Types of changes

Please put an x in the boxes related to your change.

- [ ] Bugfix
- [x] New feature or enhancement
- [ ] Others

## Checklist

*Put an x in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code.*

- [x] I have read the [CONTRIBUTING GUIDELINES](https://github.com/klaytn/klaytn/blob/master/CONTRIBUTING.md) doc
- [x] I have signed the [CLA](https://cla-assistant.io/klaytn/klaytn)
- [x] Lint and unit tests pass locally with my changes (`$ make test`)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in downstream modules

## Related issues

- Please leave the issue numbers or links related to this PR here.

## Further comments

If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by hyunsooda** [comment]
Hi,

Additional logs look good to me. I have a question. My generated file seems different from your uploaded generated `assets.go`. Did you generate it with another tool? The Klaytn has an out-of-date build recipe for `go-bindata`, but you seemed to have a different version of that tool. If so, is it well-timed to replace it together too?

**Review by hyunsooda — APPROVED** [review]

File: node/cn/tracers/testdata/call_tracer_selfdestruct.json [hunk 1]
```diff
@@ -57,7 +57,10 @@
   "result": {
     "calls": [
       {
-        "type": "SELFDESTRUCT"
+        "type": "SELFDESTRUCT",
+        "from": "0xf0b1b2a91af3b0a0a5389ea80bffdc42cf86b7e3",
+        "to": "0xf0b1b2a91af3b0a0a5389ea80bffdc42cf86b7e3",
+        "value": "0x0"
```
Comment (jeongkyun-oh) [inline node/cn/tracers/testdata/call_tracer_selfdestruct.json:63]: How about adding another testcase for non-zero value?

Comment (jeongkyun-oh) [inline node/cn/tracers/testdata/call_tracer_selfdestruct.json:63]: Is this okay without printing `gasIn` and `gasCost`?

**Review by jimni1222 — APPROVED** [review]

**Comment by aeharvlee** [comment]
> Hi,
> 
> Additional logs look good to me. I have a question. My generated file seems different from your uploaded generated `assets.go`. Did you generate it with another tool? The Klaytn has an out-of-date build recipe for `go-bindata`, but you seemed to have a different version of that tool. If so, is it well-timed to replace it together too?

I'm using `go-bindata 3.1.2 (Go runtime go1.17.8).`
```
$ go-bindata -version                                                                        
go-bindata 3.1.2 (Go runtime go1.17.8).
Copyright (c) 2010-2013, Jim Teeuwen.
```

Can you share your `go-bindata` version?
Thanks.

**Comment by hyunsooda** [comment]
@aeharvlee,
```
go-bindata 3.23.0 (Go runtime go1.16).
Copyright (c) 2010-2015, Jim Teeuwen.
Copyright (c) 2017-2022, Kevin Burke.
```

Wherefrom did you install that distro? I just installed it with `Makefile` from the Klaytn repo.

File: node/cn/tracers/testdata/call_tracer_selfdestruct.json (hunk 1 above)
Comment (aeharvlee) [reply node/cn/tracers/testdata/call_tracer_selfdestruct.json:63]: @jeongkyun-oh 
Yeap. 
`gasIn` and `gasCost` are used when [calculating `gasUsed` field of call struct](https://github.com/klaytn/klaytn/blob/dev/node/cn/tracers/internal/tracers/call_tracer.js#L164).
And both fields are not contained in the return object. You can check [this out in call_tracer.js](https://github.com/klaytn/klaytn/blob/0241baf093d0bd24217ef21664691d8b48c57461/node/cn/tracers/internal/tracers/call_tracer.js#L299-L325).
As you can see, there are no fields `gasIn` and `gasCost`.

And also the output of Ethereum's JS Call tracer after this patch is same with Klaytn.
So there will be no issue for this.

Comment (aeharvlee) [reply node/cn/tracers/testdata/call_tracer_selfdestruct.json:63]: @jeongkyun-oh 
Great. I will add the test case for non-zero value too :)

Comment (aeharvlee) [reply node/cn/tracers/testdata/call_tracer_selfdestruct.json]: @jeongkyun-oh 
I just updated the test.
balance of the contract is bigger than zero.
Because most of the scenario is same except for balance, so I chose to update the test case instead of copy and paste + update only balance field. 

Please let me know if we should keep the original test case.

**Comment by aeharvlee** [comment]
> @aeharvlee,
> 
> ```
> go-bindata 3.23.0 (Go runtime go1.16).
> Copyright (c) 2010-2015, Jim Teeuwen.
> Copyright (c) 2017-2022, Kevin Burke.
> ```
> 
> Wherefrom did you install that distro? I just installed it with `Makefile` from the Klaytn repo.

@hyunsooda 
Thanks for your info.
I installed go-bindata by using `go install` command.
I didn't know that I can get that by using `Makefile`.
I'll try it.

**Review by jeongkyun-oh — APPROVED** [review]

**Review by jimni1222 — APPROVED** [review]

**Review by hyunsooda — APPROVED** [review]
